src/Waives.Http/WaivesClient.cs

```diff
@@ -76,27 +107,49 @@ public async Task<Classifier> GetClassifier(string name)
 
         public async Task<IEnumerable<Document>> GetAllDocuments()
         {
-            var response = await HttpClient.GetAsync("/documents").ConfigureAwait(false);
+            var request = new HttpRequestMessage(HttpMethod.Get, new Uri("/documents", UriKind.Relative));
+            var response = await SendRequest(request).ConfigureAwait(false);
             await EnsureSuccessStatus(response).ConfigureAwait(false);
 
             var responseContent = await response.Content.ReadAsAsync<DocumentCollection>().ConfigureAwait(false);
-            return responseContent.Documents.Select(d => new Document(this, d.Links));
+            return responseContent.Documents.Select(d => new Document(this, d.Links, d.Id));
         }
 
         public async Task Login(string clientId, string clientSecret)
         {
-            var response = await HttpClient.PostAsync("/oauth/token", new FormUrlEncodedContent(new Dictionary<string, string>
-            {
-                { "client_id", clientId },
-                { "client_secret", clientSecret },
-            })).ConfigureAwait(false);
+            Logger.Log(LogLevel.Info, $"Logging in to Waives at {HttpClient.BaseAddress}");
 
+            var request = new HttpRequestMessage(HttpMethod.Post, new Uri("/oauth/token", UriKind.Relative))
+            {
+                Content = new FormUrlEncodedContent(new Dictionary<string, string>
+                {
+                    {"client_id", clientId},
+                    {"client_secret", clientSecret}
+                })
+            };
+
+            var response = await SendRequest(request).ConfigureAwait(false);
             await EnsureSuccessStatus(response).ConfigureAwait(false);
 
             var responseContent = await response.Content.ReadAsAsync<AccessToken>().ConfigureAwait(false);
             var accessToken = responseContent.Token;
 
             HttpClient.DefaultRequestHeaders.Add("Authorization", $"Bearer {accessToken}");
+            Logger.Log(LogLevel.Info, "Logged in.");
+        }
+
+        internal async Task<HttpResponseMessage> SendRequest(HttpRequestMessage request)
```

Comment:
:question: Would it be possible / sensible to have this as a decorator?

Reply:
Or maybe have a `LoggingHttpClient`, and have `WaivesClient` depend on a custom `IHttpClient`.